README.md
# :fire: nsw-rfs-geojson-feeds

The NSW Rural Fire Service (RFS) publishes a [GeoJSON feed of major incidents](http://www.rfs.nsw.gov.au/news-and-media/stay-up-to-date/feeds) and a [JSON feed of hazard reduction burns](https://www.rfs.nsw.gov.au/funnelback/hr-map-data?collection=nsw-rfs-hazard-xml-new), this project aims to make these feed more developer friendly.

_NSW RFS Current Incidents and Hazard Reduction data is © State of New South Wales (NSW Rural Fire Service). For current information go to www.rfs.nsw.gov.au. Licensed under the Creative Commons Attribution 4.0 International (CC BY 4.0)._

## Where is it used?

This project is used at [www.beyondtracks.com](https://www.beyondtracks.com) to provide information about bushfires and hazard reduction burns, appearing as alerts for affected walks and as [fire areas and markers on the map](https://www.beyondtracks.com/map/).

## Usage as a library

```js

const NSWRFSHazardReduction = require('@beyondtracks/nsw-rfs-geojson-feeds').NSWRFSHazardReduction
const NSWRFSMajorIncidentsGeoJSON = require('@beyondtracks/nsw-rfs-geojson-feeds').NSWRFSMajorIncidentsGeoJSON

const geojson = NSWRFSHazardReduction.toGeoJSON(feed)
const geojson = NSWRFSMajorIncidentsGeoJSON.clean(feed, {
    avoidGeometryCollections: true,
    avoidSlivers: true,
    sort: 'original'
})
```

## Usage as a CLI script

Either install with yarn or npm.

```sh
    yarn global add @beyondtracks/nsw-rfs-geojson-feeds
```

```sh
    npm install -g @beyondtracks/nsw-rfs-geojson-feeds
```

Then run the command line programs with:

```sh
    nsw-rfs-majorincidents-geojson nsw-rfs-majorincidents.geojson
    nsw-rfs-hazardreduction-geojson nsw-rfs-hazardreduction.geojson
```

This will download the upstream feeds, process them and save the resulting GeoJSON files.

To combine the two GeoJSON files into a single one you could use either of

```sh
    jq --compact-output --slurp '{type: "FeatureCollection", features: (.[0].features + .[1].features)}' nsw-rfs-majorincidents.geojson nsw-rfs-hazardreduction.geojson > nsw-rfs.geojson
```

```sh
    ogr2ogr nsw-rfs.geojson nsw-rfs-majorincidents.geojson nsw-rfs-hazardreduction.geojson
```

Alternatively if you've pre-downloaded an upstream feed and you'd like to process it you can run:

```sh
    nsw-rfs-majorincidents-geojson upstream.json output.geojson
    nsw-rfs-hazardreduction-geojson upstream.json output.geojson
```

### Options

  - `--pretty-print` pretty print the output, otherwise output is minified
  - `--avoid-geometrycollections` explode GeometryCollections out to multiple Features to avoid the use of GeometryCollections (handy for use in QGIS which doesn't support varying geometry types within a GeometryCollection, see https://github.com/qgis/QGIS/issues/32747) (off by default)
  - `--avoid-slivers` to try and remove narrow slivers in polygons (off by default)
  - `--sort=<original|guid|pubdate>` the default is `original` which retains the original sort order, sorting by `guid` uses ascending order and `pubdate` uses chronological order

## Hosted Feeds

1. [https://www.beyondtracks.com/contrib/nsw-rfs-majorincidents.geojson](https://www.beyondtracks.com/contrib/nsw-rfs-majorincidents.geojson) - [preview with geojson.io](http://geojson.io/#data=data:text/x-url,https://www.beyondtracks.com/contrib/nsw-rfs-majorincidents.geojson)
2. [https://www.beyondtracks.com/contrib/nsw-rfs-majorincidents-nogeometrycollections.geojson](https://www.beyondtracks.com/contrib/nsw-rfs-majorincidents-nogeometrycollections.geojson) - [preview with geojson.io](http://geojson.io/#data=data:text/x-url,https://www.beyondtracks.com/contrib/nsw-rfs-majorincidents-nogeometrycollections.geojson) - processed with the `--avoid-geometrycollections` option for QGIS compatibility
3. [https://www.beyondtracks.com/contrib/nsw-rfs-hazardreduction.geojson](https://www.beyondtracks.com/contrib/nsw-rfs-hazardreduction.geojson) - [preview with geojson.io](http://geojson.io/#data=data:text/x-url,https://www.beyondtracks.com/contrib/nsw-rfs-hazardreduction.geojson)

An archive of historical feed data is kept as a Git repository at [https://github.com/beyondtracks/nsw-rfs-majorincidents-archive](https://github.com/beyondtracks/nsw-rfs-majorincidents-archive) together with a time series visualisation at [https://github.com/beyondtracks/nsw-rfs-majorincidents-timeseries](https://github.com/beyondtracks/nsw-rfs-majorincidents-timeseries).

## Features
### Access-Control-Allow-Origin

The upstream feed lacks the Access-Control-Allow-Origin header which means web applications aren't able to use the feed directly. This was reported to the NSW RFS on the 3rd of December 2015, and as of December 2019 the header still isn't present.

The sample crontab file allows you to mirror the RFS feed and serve it with your own HTTP server, adding the Access-Control-Allow-Origin HTTP header.

### Nested GeometryCollections

The upstream feed uses nested GeometryCollections, although allowed by the [GeoJSON spec](https://tools.ietf.org/html/rfc7946), the recommendation is they SHOULD be avoided.

Nested GeometryCollections are converted to flat GeometryCollections when they include different `Geometry` types or a multipart geometry where it's possible to avoid using a GeometryCollection at all.

Since some downstream applications can have trouble ingesting GeoJSON data with GeometryCollections of varying types, the option `--avoid-geometrycollections` is provided to explode these out into multiple Features.

### Coordinate Precision

Although extra coordinate precision can help retain geometry shape even beyond the capture precision, the upstream feed uses 14 decimal places, which is more than reasonable precision for Point features. Point geometry coordinates are limited to 4 decimal places for ~10m resolution, polygon geometry coordinates are unchanged to prevent creating invalid polygons.

### Overloaded Description

The upstream feed overloads properties into the `description` field in the format `KEY: Value <br />KEY: Value`. These are exploded out to make them easier to machine read in applications. The original overloaded description is dropped from the output.

Some properties are parsed:

- fire size is converted from a string "10 ha" into a numeric value with units in hectares

### Machine Readable Schema

Within [schema/](https://github.com/beyondtracks/nsw-rfs-majorincidents-geojson/tree/master/schema) are JSON files containing the values and descriptions for Status, Alert Level and Incident Type. These can be used within web applications to provide users more information about what these terms mean.

### ISO8601 Datetimes

The upstream feed uses dates in the format `3/01/2018 5:20:00 AM` in some places and `3 Jan 2018 16:20` in other places in local time. These datetimes are converted into ISO8601 datetimes assuming the 'Australia/Sydney' time zone to avoid any ambiguities in interpretation.

### Winding Order

For extra assurances the GeoJSON winding order is enforced with https://github.com/mapbox/geojson-rewind.

### Sorted features

You can control the order of features with `--sort=original|guid|pubdate`. The default is `original` which retains the original sort order, sorting by `guid` uses ascending order and `pubdate` uses chronological order.

### Removal of Internal Shared Borders
Around November 2019 it was observed some bushfire areas were being split into multiple `Polygon` geometries within the `GeometryCollection` for the incident. Since these are purely artificial, there is no compelling reason to include them, so we attempt to remove these by unioning multiple Polygons together with [polygon-clipping](https://github.com/mfogel/polygon-clipping).

![Shared internal borders](img/shared-borders.png)

Where the internal shared borders aren't perfectly touching you can try with the `--avoid-slivers` option (off by default) which tries to remove slivers less than 25m in width.

### ID

The `guid` is used as a numeric GeoJSON Feature ID.


## Schema
Each GeoJSON Feature represents a major incident. Each Feature may have multiple geometries using a GeometryCollection. For example both a point indicating a rough location and a polygon showing a bushfire extent. Although so far every feature contains at a minimum a point geometry, without any documented guarantees about this consumers should accept features with no geometry or only a polygon geometry.

Each feature has a properties field as follows. Note that these vary from the upstream feed, the full list of possible values is unknown these are just the ones I've come across. None of these properties are guaranteed to be present.

Key   | Type | Description | Example
------|------|-------------|---------
title | String | A short name used as the tile of this incident, given to assist with managing multiple incidents. Usually this is the name of a nearby road or geographic feature but it doesn't necessarily reflect the exact location of that incident. | `Grain Valley Rd, Boggabri`
link | String | A URL where more information about the incident can be found. If not found the generic link `http://www.rfs.nsw.gov.au/fire-information/fires-near-me` should be used. | `http://www.rfs.nsw.gov.au/fire-information/major-fire-updates/mfu?id=6025`
link-updated | String | Datetime in ISO8601 for the when the link contents were last updated. | `2019-12-21T21:14:00+11:00`
alert-level | String | Alert level of the incident, see [#alert-level](https://github.com/beyondtracks/nsw-rfs-majorincidents-geojson#alert-level) | `emergency-warning`, `watch-and-act`, `advice`, `not-applicable`
status | String | Status of the incident, see [#status](https://github.com/beyondtracks/nsw-rfs-majorincidents-geojson#status) | `out-of-control`, `being-controlled`, `under-control`
guid | String | A globally unique identifier for this incident. Usually it is a non-publicly accessible URL | `https://incidents.rfs.nsw.gov.au/api/v1/incidents/285935`
pub-date | String | Datetime in ISO8601 of when the incident was first published | `2018-01-05T23:41:00+11:00`
updated | String | Datetime in ISO8601 of when the incident details were last updated | `2018-01-06T10:41:00+11:00`
responsible-agency | String | The name of the agency responsible for responding to the incident | `Rural Fire Service`, `NSW National Parks and Wildlife Service`
size | String | The ground area which the incident affects in hectares | `2790`
type | String | The type of incident, see [#incident-type](https://github.com/beyondtracks/nsw-rfs-majorincidents-geojson#incident-type) | `bush-fire`, `medical`, `grass-fire`
fire | boolean | Indicates if this this incident a fire (`true`) or not (`false`) | `true`, `false`
council-area | String | Short name of the council area the incident falls in | `Wollongong`
location | String | A description of the location of the incident | `Coast Trk, Lilyvale, NSW 2508`, `150.79 -33.30`

### Status

Status Name | Description
------------|-------------
Out of control | A fire which is spreading on one or more fronts. Effective containment strategies are not in place for the entire perimeter.
Being controlled | Effective strategies are in operation or planned for the entire perimeter.
Under control | The fire is at a stage where fire fighting resources are only required for patrol purposes and major re-ignition is unlikely.

### Alert Level

Alert Level Name | Description
-----------------|-------------
Emergency Warning| An Emergency Warning is the highest level of Bush Fire Alert. You may be in danger and need to take action immediately. Any delay now puts your life at risk.
Watch and Act    | There is a heightened level of threat. Conditions are changing and you need to start taking action now to protect you and your family.
Advice           | A fire has started. There is no immediate danger. Stay up to date in case the situation changes.
Not Applicable   | |

### Incident Type

Incident Type Name | Description
-------------------|-------------
Bush Fire | Forest and/or Scrub Fire
Grass Fire | Grass Fire
Hazard Reduction | Planned controlled burns to reduce bush fire hazards
Structure Fire | A fire involving a residential, commercial or industrial building
Haystack Fire | Haystack fire
HAZMAT | The NSW RFS provides operational support to the Fire & Rescue NSW for hazardous materials incidents
MVA/Transport | Transport incidents including motor vehicle accident, aircraft incident and incidents involving a railway or railway rolling stock
Assist Other Agency | Assist other agency such as Fire & Rescue NSW, NSW Police, NSW SES, NSW Ambulance, Defence Force, interstate deployments etc
Search/Rescue | Search and Rescue, rescue animal
Flood/Storm/Tree Down | Assisting with Flood or storm damage or tree down
Vehicle/Equipment Fire | Vehicle or Equipment fire such as car fire, farm machinery fire, transformer fire etc
Burn off | Burn off such as stubble fire, pile burn etc
Fire Alarm | Automatic Fire Alarm or Domestic Smoke Alarm
Medical | Medical incident such as medical evacuation
Other | Other incidents such as smoke in vicinity, gas leak, building collapse etc

## Warranty

The information in the RFS feed can affect life and property. Although the aim of this project is to make the RFS feed more safe, usable and reliable for data consumers, errors or omissions may be present and/or the upstream supplied data structure may change without any notice causing issues. Use at your own risk.

THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.


CHANGELOG.md
# 4.0.1

* Package installs executable scripts: `nsw-rfs-hazardreduction-geojson`, `nsw-rfs-majorincidents-geojson`

# 4.0.0

* Package renamed to `@beyondtracks/nsw-rfs-geojson-feeds`.

# 3.0.0

* Add processing for hazard reduction feed

# 2.7.0

* Limit coordinate precision of points
* Parse size property as a number (hectares)

# 2.6.0

* Attempt to parse out the major fire update link and insert it as the link as this is more useful. This also adds a new property 'link-updated` which contains the time contained in the "MAJOR FIRE UPDATE AS AT...".

# 2.5.0

* `--sort=original|guid|pubdate` option to sort features.
* `guid` is promoted to a GeoJSON Feature `id`.

# 2.4.1

* Fix an issue in description parsing where the key contained a `:`, "MAJOR FIRE UPDATE AS AT...".

# 2.4.0

* `--avoid-geometrycollections` to explode GeometryCollections out into mulitple Features.
* `--avoid-slivers` to remove slivers in Polygons.
* Update @mapbox/geojson-rewind to fix issue rewinding Polygons within GeometryCollections.
* Try to avoid internal borders by unioning Polygons.
* Use eslint for code linting.

# 2.3.0

* `--pretty-print` option added.

# 2.2.1

* Fix output to stdout support.

# 2.2.0

* Drop polygons with 0 area
* Support pre-downloaded input
* Disable coordinate precision limit since it created invalid polygons

# 2.1.1

* Fix schema trailing space

# 2.1.0

* Remove titles from schema
* Sort features

# 2.0.0

* Drop identifiers and include JSON schema
* Changes to the properties schema




e2eTest.sh
#!/bin/sh

mkdir -p output

echo "Test 1: download and process major incidents"
./bin/nsw-rfs-majorincidents-geojson --pretty-print output/nsw-rfs-majorincidents.geojson
geojsonhint output/nsw-rfs-majorincidents.geojson

echo "Test 2: download and process major incidents --avoid-geometrycollections"
./bin/nsw-rfs-majorincidents-geojson --pretty-print --avoid-geometrycollections output/nsw-rfs-majorincidents-avoidgeometrycollections.geojson
geojsonhint output/nsw-rfs-majorincidents-avoidgeometrycollections.geojson

echo "Test 3: download and process major incidents --avoid-slivers"
./bin/nsw-rfs-majorincidents-geojson --pretty-print --avoid-slivers output/nsw-rfs-majorincidents-avoidslivers.geojson
geojsonhint output/nsw-rfs-majorincidents-avoidslivers.geojson

echo "Test 4: download and process hazard reduction"
./bin/nsw-rfs-hazardreduction-geojson --pretty-print output/nsw-rfs-hazardreduction.geojson
geojsonhint output/nsw-rfs-hazardreduction.geojson

echo "Test 5: input fixture hazard reduction"
./bin/nsw-rfs-hazardreduction-geojson --pretty-print test/fixtures/input/hr.json - | diff test/fixtures/output/hr.geojson -



src/hazardreduction.js
const { DateTime } = require('luxon')
const rewind = require('@mapbox/geojson-rewind')

module.exports = {
  /**
   * Converts the NSW RFS Hazard Reduction JSON feed to GeoJSON
   *
   * @param {Object} json NSW RFS Hazard Reduction Feed as a JSON Object
   * @returns {Object} GeoJSON Object
   */
  toGeoJSON(json) {
    const self = this

    const features = json.results.map(result => {
      const polygons = result.polygons.map(polygon => {
        return [polygon
          .polygon
          .split('|')
          .map(coordinatePair => {
            return coordinatePair
              .split(';')
              .filter(coordinates => coordinates.length >= 2)
              .map(coordinate => Number(coordinate))
              .reverse()
          })
          .filter(coordinates => coordinates.length)]
      })

      return rewind({
        type: 'Feature',
        id: result.guarReference,
        properties: {
          hrNotice: true, // flag this as a hazard reduction notice so that when combined with current incidents we can tell apart planned from actual
          leadAgency: result.leadAgency,
          supportingAgencies: result.supportingAgencies,
          size: self._cleanSize(result.size),
          title: result.location.replace(/\s*HAZARD REDUCTION\s*$/i, ''), // remove trailing HAZARD REDUCTION from title
          tenure: result.tenure,
          startDate: self._cleanDate(result.startDate),
          endDate: self._cleanDate(result.endDate)
        },
        geometry: {
          type: result.geometryType,
          coordinates: polygons
        }
      })
    })

    return {
      type: 'FeatureCollection',
      features: features
    }
  },

  /*
   * Given an date string like "18/10/2020" return as ISO8601 assuming local time zone of Australia/Sydney
   *
   * @param {String} date A date string in "18/10/2020" format
   * @returns {String} ISO8601 date
   */
  _cleanDate(date) {
    return DateTime.fromFormat(date, 'd/MM/yyyy', { zone: 'Australia/Sydney' }).toISODate()
  },

  /*
   * Given an string like "1.2 ha" return size as a number assuming units in ha
   *
   * @param {String} datetime A size in "1.2 ha" format
   * @returns {number} size as a number
   */
  _cleanSize(size) {
    if (typeof size === 'string') {
      const area = Number(size.replace(/\s*ha\s*$/i, ''))
      if (isNaN(area)) {
        return null
      } else {
        return area
      }
    } else {
      return size
    }
  }
}


src/majorincidents.js
const turf = {
  area: require('@turf/area').default,
  featureEach: require('@turf/meta').featureEach,
  feature: require('@turf/helpers').feature,
  featureCollection: require('@turf/helpers').featureCollection,
  buffer: require('@turf/buffer')
}
const moment = require('moment-timezone')
const rewind = require('@mapbox/geojson-rewind')
const _ = require('lodash')
const polygonClipping = require('polygon-clipping')
const gp = require('geojson-precision')

const defaultOptions = {
  avoidGeometryCollections: false,
  avoidSlivers: false
}

module.exports = {
  /**
     * "Cleans" the NSW RFS Major Incidents Upstream Feed to:
     *
     *   - Avoid nested GeometryCollections
     *   - Limit coordinate precision
     *   - Unpack overloaded description field
     *   - Use ISO8601 formatted datetimes
     *   - Enforce GeoJSON winding order
     *   - Union mulitple Polygons within a single GeometryCollection due to artificial shared borders
     *   - Covert size from a string to a number (in hectares)
     *
     * @param {Object} geojson NSW RFS Major Incidents Upstream Feed as a GeoJSON Object
     * @param {Object} [options]
     * @param {boolean} [options.avoidGeometryCollections=false] Avoid GeometryCollections and explode into flat Features (this may increase the feature count and duplicate properties across multiple features). Defaults to false.
     * @param {boolean} [options.avoidSlivers=false] Try to avoid slivers in the Polygons. Defaults to false.
     * @param {original|guid|pubdate} [options.sort=original] Sort features by. Defaults to `original`.
     * @returns {Object} The "cleaned" GeoJSON Object
     */
  clean(geojson, options) {
    const self = this

    options = Object.assign({}, defaultOptions, options)

    // clean up the upstream GeoJSON
    let cleanFeatures = []
    turf.featureEach(geojson, (feature) => {
      const cleanGeometry = self._cleanGeometry(feature.geometry, {avoidSlivers: options.avoidSlivers})
      const cleanProperties = self._cleanProperties(feature.properties)

      if (options.avoidGeometryCollections) {
        if (cleanGeometry.type === 'GeometryCollection') {
          cleanGeometry.geometries.forEach((geometry) => {
            const cleanFeature = turf.feature(geometry, cleanProperties)
            cleanFeatures.push(cleanFeature)
          })
        }
      } else {
        const cleanFeature = turf.feature(cleanGeometry, cleanProperties)
        cleanFeatures.push(cleanFeature)
      }
    })

    // promote GUID to id
    cleanFeatures = cleanFeatures.map((feature) => {
      const featureID = self._extractID(feature.properties.guid)
      if (featureID !== null) {
        feature.id = featureID
      }
      return feature
    })

    // sort features
    const sortBy = {
      guid: 'id',
      pubdate: i => (new Date(i.properties['pub-date'])).getTime()
    }
    if (options.sort && options.sort in sortBy) {
      cleanFeatures = _.sortBy(cleanFeatures, [sortBy[options.sort]])
    }

    // create final GeoJSON with winding order enforced, and
    // limited coordinate precision for points (polygons are left in tact to avoid
    // creating invalid geometries)
    const cleanedGeoJSON = gp(rewind(turf.featureCollection(cleanFeatures)), 4, 4, {
      skipLineString: true,
      skipPolygon: true
    })

    return cleanedGeoJSON
  },

  _sortIndexStatus(value) {
    switch (value) {
    case 'Out of control':
      return 0
    case 'Being controlled':
      return 1
    case 'Under control':
      return 3
    default:
      return 4
    }
  },

  _sortIndexAlertLevel(value) {
    switch (value) {
    case 'Emergency Warning':
      return 0
    case 'Watch and Act':
      return 1
    case 'Advice':
      return 2
    case 'Not Applicable':
      return 3
    default:
      return 4
    }
  },

  /**
     * Given a GUID string in the form `https://incidents.rfs.nsw.gov.au/api/v1/incidents/361069`
     * return the ID as a number.
     *
     * @param {string} guid
     * @returns {Number}
     * @private
     */
  _extractID(guid) {
    if (guid) {
      const matches = guid.match(/([0-9]*)$/)
      if (matches && matches.length === 2) {
        const integer = Number.parseInt(matches[1])
        if (Number.isFinite(integer)) {
          return integer
        }
      }
    }

    return null
  },

  /**
     * Given a GeoJSON Geometry object, flattens any GeometryCollection's out
     * into an Array of Geometry objects.
     *
     * @param {Object} geometry A GeoJSON Geometry object
     * @returns {Array} An Array of GeoJSON Geometry objects
     * @private
     */
  _flattenGeometries(geometry) {
    const self = this

    if (geometry === undefined) {
      return []
    }

    if (geometry === null) {
      return [geometry]
    }

    if (geometry.type !== 'GeometryCollection') {
      return [geometry]
    } else if (geometry.geometries && geometry.geometries.length) {
      return _.flattenDeep(geometry.geometries.map(g => self._flattenGeometries(g)))
    } else {
      return []
    }
  },

  /**
     * Given an Array of GeoJSON GeometryCollection geometries determines if they
     * are all of uniform type.
     *
     * @param {Array} geometryList An Array of GeoJSON GeometryCollection geometries
     * @returns {boolean}
     * @private
     */
  _uniformType(geometryList) {
    if (geometryList.length < 2) {
      return true
    } else {
      return geometryList.map(geom => geom.type).reduce((acc, cur) => acc === cur)
    }
  },

  /**
     * Avoids nested GeometryCollections where possible and converts single type
     * GeometryCollections into multipart geometry types. Where the GeometryCollection
     * contains mixed geometries, they are also reduced into a GeometryCollection of
     * multipart geometries.
     *
     * @param {Object} geometry A GeoJSON Geometry
     * @params {Object} [options]
     * @params {boolean} [options.avoidSlivers=false] Try to avoid slivers in the Geometry by buffering then reverse buffering.
     * @returns {Object} A GeoJSON Geometry avoiding nested GeometryCollections and using multipart geometry types in favour of single type GeometryCollections
     * @private
     */
  _cleanGeometry(geometry, options) {
    options = Object.assign({}, {avoidSlivers: false}, options)

    // explode GeometryCollections into an array of Geometries
    // also removing any 0 area polygons
    const flatGeometries = this._flattenGeometries(geometry)
      .map((g) => {
        if (g && g.type === 'Polygon' && turf.area(g) === 0) {
          // not a valid polygon
          return null
        }
        return g
      })
      .filter(g => g !== null)

    if (!flatGeometries.length) return null

    if (flatGeometries.length === 1) {
      // a single geomtery can be returned as is
      return flatGeometries[0]
    } else {
      // Attempt to union multiple Polygons found within the GeometryCollection,
      // since as of Dec 2019 large polygons are being split on artificial internal
      // boundaries which we try to remove.
      const sliverBuffer = 25
      const polygons = flatGeometries.filter(geometry => geometry.type === 'Polygon').map(geometry => (options.avoidSlivers ? turf.buffer(geometry, sliverBuffer, {units: 'meters'}).geometry : geometry))
      const nonPolygons = flatGeometries.filter(geometry => geometry.type !== 'Polygon')

      const flatGeometriesUnioned = nonPolygons
      if (polygons.length) {
        try {
          let unioned
          if (polygons.length > 1) {
            const polygonsUnioned = polygonClipping.union(...polygons.map(g => g.coordinates))
            if (polygonsUnioned.length > 1) {
              unioned = {
                type: 'MultiPolygon',
                coordinates: polygonsUnioned
              }
            } else {
              unioned = {
                type: 'Polygon',
                coordinates: polygonsUnioned[0]
              }
            }
          } else {
            unioned = polygons[0]
          }

          if (options.avoidSlivers) {
            flatGeometriesUnioned.push(turf.buffer(unioned, -sliverBuffer, {units: 'meters'}).geometry)
          } else {
            flatGeometriesUnioned.push(unioned)
          }
        } catch (e) {
          // if there was an error unioning polygons, then still output them in their original form
          console.error('Union error', e)
          if (options.avoidSlivers) {
            flatGeometriesUnioned.push(...(turf.buffer(polygons, -sliverBuffer, {units: 'meters'})))
          } else {
            flatGeometriesUnioned.push(...polygons)
          }
        }
      }

      if (this._uniformType(flatGeometriesUnioned)) {
        const type = flatGeometriesUnioned[0].type

        // can be converted into a multi geom type
        if (type === 'MultiPolygon') {
          return flatGeometriesUnioned[0]
        } else {
          return {
            type: `Multi${  type}`,
            coordinates: flatGeometriesUnioned.map(g => g.coordinates)
          }
        }
      } else {
        // can't be converted into a geom type, use GeometryCollection instead
        // however still check if any of the geometries can be combined into multi types
        const points = flatGeometriesUnioned.filter(g => g.type === 'Point')
        const lines = flatGeometriesUnioned.filter(g => g.type === 'LineString')
        const polygons = flatGeometriesUnioned.filter(g => g.type === 'Polygon')
        const others = flatGeometriesUnioned.filter(g => !(['Point', 'LineString', 'Polygon'].includes(g.type)))

        const geometries = []
        if (points.length > 1) {
          geometries.push({
            type: 'MultiPoint',
            coordinates: points.map(g => g.coordinates)
          })
        } else if (points.length === 1) {
          geometries.push(points[0])
        }

        if (lines.length > 1) {
          geometries.push({
            type: 'MultiLineString',
            coordinates: lines.map(g => g.coordinates)
          })
        } else if (lines.length === 1) {
          geometries.push(lines[0])
        }

        if (polygons.length > 1) {
          geometries.push({
            type: 'MultiPolygon',
            coordinates: polygons.map(g => g.coordinates)
          })
        } else if (polygons.length === 1) {
          geometries.push(polygons[0])
        }

        if (others.length > 1) {
          geometries.push(...others)
        } else if (others.length === 1) {
          geometries.push(others[0])
        }

        return {
          type: 'GeometryCollection',
          geometries
        }
      }
    }
  },

  _cleanProperties(properties) {
    if (properties.pubDate) {
      properties.pubDate = this._cleanPubDate(properties.pubDate)
      properties['pub-date'] = properties.pubDate
      delete properties.pubDate
    }

    if (properties.description) {
      Object.assign(properties, this._unpackDescription(properties.description))
      delete properties.description // remove original description string
    }

    if (properties.size) {
      if (typeof properties.size === 'string') {
        const area = Number(properties.size.replace(/\s*ha\s*$/i, ''))
        if (isNaN(area)) {
          delete properties.size
        } else {
          properties.size = area
        }
      }
    }

    /* use a simplified schema removing some keys and using identifiers rather than full names for some values */

    // ALERT LEVEL seems to be a duplicate of category, best practice is to avoid duplication to reduce the risk of them coming out of sync
    if ('alert-level' in properties && 'category' in properties && (properties.category !== properties['alert-level'])) {
      console.error('properties.category !== description.ALERT LEVEL!', properties.category, properties['alert-level'])
    }

    if ('alert-level' in properties)
      delete properties['alert-level']

    if ('category' in properties) {
      properties['alert-level'] = properties.category
      delete properties.category
    }

    if ('guid_isPermaLink' in properties)
      delete properties['guid_isPermaLink']

    if ('fire' in properties)
      properties.fire = !!properties.fire.match(/Yes/i)

    // since this is a generic link applying to every incident don't bother to include it for each
    if ('link' in properties && properties.link === 'http://www.rfs.nsw.gov.au/fire-information/fires-near-me') {
      delete properties.link
    }

    return properties
  },


  /**
     * Given an date string like "3/01/2018 5:20:00 AM" as used in the pubDate
     * field, return in ISO8601 format. Assumes time in local times of
     * Australia/Sydney.
     *
     * @param {String} datetime A date string like "3/01/2018 5:20:00 AM" as used in the pubDate field
     * @returns {String} An ISO8601 formatted datetime
     * @private
     */
  _cleanPubDate(datetime) {
    return moment.tz(datetime, 'D/MM/YYYY h:mm:ss A', 'Australia/Sydney').format()
  },

  /**
     * Given an date string like "3 Jan 2018 16:20" as used in the UPDATED field
     * within the description, return in ISO8601 format. Assumes time in local
     * times of Australia/Sydney.
     *
     * @param {String} datetime A date string like "3 Jan 2018 16:20" as used in the UPDATED field
     * @param {boolean} twelveHour `true` if the date is in 12 hour format, otherwise assumes 24 hour format.
     * @returns {String} An ISO8601 formatted datetime
     * @private
     */
  _cleanUpdatedDate(datetime, twelveHour) {
    return moment.tz(datetime, twelveHour ? 'D MMM YYYY HH:mmA' : 'D MMM YYYY HH:mm', 'Australia/Sydney').format()
  },

  /**
     * Unpacks the overloaded description string like "KEY1: Value1 <br />KEY 2: Value2"
     * into an Object like:
     *
     *    {
     *       "key": "Value1",
     *       "key-2": "Value2"
     *    }
     *
     *  Keys are lower cased and have spaces replaced with a dash character `-`.
     *
     * @param {String} description The description field string.
     * @returns {Object} An Object of the unpacked description
     * @private
     */
  _unpackDescription(description) {
    const self = this

    if (!description)
      return {}

    const lines = description.split(/ *<br ?\/?> */)
    const result = {}
    for (const line of lines) {
      const splits = line.split(':')
      if (splits && splits.length >= 2) {
        let key
        let value
        // when they use MAJOR FIRE UPDATE AS AT 15 Dec 2019 11:12PM: http://
        // they must have really wanted to make life harder for us and made
        // applications more likely to break without warning
        if (splits[0].startsWith('MAJOR FIRE UPDATE AS AT')) {
          // take the first two as the key and use the rest as value
          const keyParts = []
          keyParts.push(splits.shift())
          keyParts.push(splits.shift())

          key = keyParts.join(':')
          value = splits.join(':').trim()

          // now translate
          //   key: MAJOR FIRE UPDATE AS AT 21 Dec 2019 11:12PM
          //   value: <a href='http://www.rfs.nsw.gov.au/fire-information/major-fire-updates/mfu?id=6025' target='_blank'> More information</a>
          // into
          //   link: http://www.rfs.nsw.gov.au/fire-information/major-fire-updates/mfu?id=6025
          //   link-updated: 2019-12-21T23:12:00+11:00
          const asAt = key.replace(/^MAJOR FIRE UPDATE AS AT /, '')
          const date = this._cleanUpdatedDate(asAt, true)
          const matches = value.match('(http[^\']*)')
          if (matches && matches.length === 2) {
            const link = matches[1]

            result['major-fire-update-link'] = link
            result['major-fire-update-updated'] = date
            continue
          }
        } else {
          // take the first one as the key
          key = splits.shift()
          // join the rest into the value
          value = splits.join(':').trim()
        }

        if (key === 'UPDATED') {
          value = self._cleanUpdatedDate(value)
        }

        // lower case keys and use - instead of spaces
        key = key.replace(/ /g, '-').toLowerCase()

        result[key] = value
      }
    }

    if ('major-fire-update-link' in result) {
      result.link = result['major-fire-update-link']
      delete result['major-fire-update-link']
    }
    if ('major-fire-update-updated' in result) {
      result['link-updated'] = result['major-fire-update-updated']
      delete result['major-fire-update-updated']
    }

    return result
  }
}
